backend/crud.py

```diff
@@ -0,0 +1,211 @@
+from typing import Dict
+from sqlalchemy.orm import Session
+from sqlalchemy import select, update
+
+from .models import (
+    AttrType,
+    Attribute,
+    AttributeDefinition,
+    BoundFK,
+    Entity,
+    Schema,
+    Value,
+    val_map,
+    type_map
+)
+
+from .schemas import (
+    SchemaCreateSchema,
+    SchemaUpdateSchema,
+    AttributeCreateSchema
+)
+
+
+def create_attribute(db: Session, data: AttributeCreateSchema) -> Attribute:
+    exists = db.execute(select(Attribute).where(Attribute.name == data.name)).scalar()
+    if exists:
+        # TODO custom exceptions
+        raise Exception(f'Attribute with name `{data.name}` already exists')
```

Comment:
As you suspected in the chat, this might be a candidate for _silent failure_.

I assume you had an alternative in mind like the `get_or_create()` method of Django QuerySets. That would indeed make sense to me.

Some additional thoughts:

* If this function gets exposed via the API, it would be nice to indicate the difference (attribute created or already existing) via different HTTP status codes (e.g. 201 and 200).
* We might hide this function even from clients. If a client requests the creation of an attribute definition (as part of a schema definition), we might already offer the name and type of the attribute instead of the attribute ID.

Reply:
In a Django project I ran into problems with simultaneous calls to `get_or_create` methods. As a mere precaution I would suggest to use a lock here.

Reply:
> I assume you had an alternative in mind like the get_or_create() 

Actually I haven't thought about it😅 But yeah, it might be useful to go this way

> indicate the difference via different HTTP status codes

Makes sense

> would suggest to use a lock here

Makes sense

> We might hide this function even from clients

I don't think that this is possible — attributes have to appear from somewhere, right?

> If a client requests the creation of an attribute definition we might already offer the name and type of the attribute

That's actually how I was thinking it should work: firstly users define the attribute itself and when they create schema they can choose from list of existing attributes

Reply:
Yes, that is one approach. When writing the comment, I was thinking that it would be nice for clients, if they would not need to deal with the attribute creation directly. I imaged a request with data like this for schema creation:

```json
{
  "name": "New Schema",
  "slug": "new-schema",
  "attributes": [
    {
      "name": "foo",
      "type": "str",
      "required": true
    },
    {
      "name": "bar",
      "type": "float",
      "list": true,
      "required": false
    }
  ]
}
```

So, during the schema creation, the backend would take care of creating the attributes as needed. So one the frontend the attributes can be defined freely.

What do you think? Can that work for you?

Reply:
Well, that will complicate things, but yes, that's possible